File: LargestValueAtEachLevel.java
// Did this code successfully run on Leetcode : Yes 
// Any problem you faced while coding this : No

//Time Complexity : O(n) since we are iterating through all the levels and nodes
//Space Complexity: O(1) since we are not using extra space
class Solution {
    public List<Integer> largestValues(TreeNode root) {
          
        List<Integer> returnList = new ArrayList<>();
        
        if(root == null) return returnList;
        
        Queue<TreeNode> treeQueue = new LinkedList<>();
        treeQueue.add(root);
        while(!treeQueue.isEmpty()){
            int size = treeQueue.size();

            int max = Integer.MIN_VALUE;
            for(int i = 0; i < size; i++){
                TreeNode curr = treeQueue.poll();
                if(curr.left != null){
                    treeQueue.add(curr.left);
                }
                
                if(curr.right != null){
                   treeQueue.add(curr.right);
                }                
                max = Math.max(max, curr.val); //Calculate max value at each level
            }           
            returnList.add(max);
        }
        
        return returnList;
    }
    
}

File: largestValueAtEachLevel.js
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes

// Any problem you faced while coding this : no

// BFS Approach
var largestValues = function (root) {
  let que = [root];
  const resultsArr = [];
  if (root === null) return resultsArr;

  while (que.length) {
    let len = que.length;
    let compareArr = [];

    while (len--) {
      root = que.shift();
      if (root.left !== null) {
        que.push(root.left);
      }
      if (root.right !== null) {
        que.push(root.right);
      }
      compareArr.push(root.val);
    }
    let greatestVal = Math.max(...compareArr);
    resultsArr.push(greatestVal);
  }
  return resultsArr;
};
